File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import qualified Data.Char as C
import Data.Text (Text)
import qualified Data.Text as T

responseFor :: Text -> Text
responseFor xs
  | invalid = T.pack "Fine. Be that way!"
  | yelling && question = T.pack "Calm down, I know what I'm doing!"
  | question = T.pack "Sure."
  | yelling = T.pack "Whoa, chill out!"
  | otherwise = T.pack "Whatever."
  where
    stripped = T.strip xs
    yelling = not (T.any C.isLower stripped) && T.any C.isLetter stripped
    question = T.last stripped == '?'
    invalid = not (T.any (\x -> C.isAlphaNum x || C.isPunctuation x) stripped)

File: python/bob/bob.py
def response(hey_bob: str):
    hey_bob = hey_bob.strip()
    if hey_bob == "":
        return "Fine. Be that way!"
    elif hey_bob.isupper():
        if hey_bob[-1] == "?":
            return "Calm down, I know what I'm doing!"
        else:
            return "Whoa, chill out!"
    elif hey_bob[-1] == "?":
        return "Sure."
    else:
        return "Whatever."
